features/lorawan/LoRaWANInterface.cpp

```diff
@@ -0,0 +1,177 @@
+/**
+ * @file
+ *
+ * @brief      Implementation of LoRaWANBase
+ *
+ * Copyright (c) 2017, Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "lorawan/LoRaWANInterface.h"
+
+inline LoRaWANStack& stk_obj()
+{
+    return LoRaWANStack::get_lorawan_stack();
+}
+
+LoRaWANInterface::LoRaWANInterface(LoRaRadio& radio)
+{
+    // Pass mac_handlers to radio to the radio driver after
+    // binding radio driver to PHY layer
+    radio_events_t *events = stk_obj().bind_radio_driver(radio);
+    radio.lock();
+    radio.init_radio(events);
+    radio.unlock();
+}
+
+LoRaWANInterface::~LoRaWANInterface()
+{
+}
+
+lora_mac_status_t LoRaWANInterface::initialize()
+{
+    return stk_obj().initialize_mac_layer();
+}
+
+lora_mac_status_t LoRaWANInterface::connect()
+{
+    // connection attempt without parameters.
+    // System tries to look for configuration in mbed_lib.json that can be
+    // overridden by mbed_app.json. However, if none of the json files are
+    // available (highly unlikely), we still fallback to some default parameters.
+    // Check lorawan_data_structure for fallback defaults.
+
+    lorawan_connect_t connection_params;
+
+    if (OVER_THE_AIR_ACTIVATION != 0) {
+        static uint8_t dev_eui[] = LORAWAN_DEVICE_EUI;
+        static uint8_t app_eui[] = LORAWAN_APPLICATION_EUI;
+        static uint8_t app_key[] = LORAWAN_APPLICATION_KEY;
+        /**
+         *
+         * OTAA join
+         */
+        connection_params.connect_type = LORAWAN_CONNECTION_OTAA;
+        connection_params.connection_u.otaa.app_eui = app_eui;
+        connection_params.connection_u.otaa.dev_eui = dev_eui;
+        connection_params.connection_u.otaa.app_key = app_key;
+        connection_params.connection_u.otaa.nb_trials = LORAWAN_NB_TRIALS;
+
+        return connect(connection_params);
+    } else {
+        static uint8_t nwk_skey[] = LORAWAN_NWKSKEY;
+        static uint8_t app_skey[] = LORAWAN_APPSKEY;
+        static uint32_t dev_addr = LORAWAN_DEVICE_ADDRESS;
+        static uint32_t nwk_id = (LORAWAN_DEVICE_ADDRESS & LORAWAN_NETWORK_ID_MASK);
```

Comment:
Is this only used in one place? If so can this be moved from the header file into local scope?

Reply:
Yeah you are right. I will look into it.

Reply:
Actually the macros are used elsewhere. And there are other macros as well which are configurable via json and for consistency they should remain together. If I move these ones out, its not consistent anymore. All configurable parameters are defined in lora_mac_data_structures.h